control-plane/connect-inject/peering_dialer_controller_test.go

```diff
@@ -18,6 +23,465 @@ import (
+				//require.Equal(t, api.PeeringStateActive, peering.State)
+				require.NotEmpty(t, peering.ID)
+
+				// Get the reconciled PeeringDialer and make assertions on the status
+				dialer := &v1alpha1.PeeringDialer{}
+				err = fakeClient.Get(context.Background(), namespacedName, dialer)
+				require.NoError(t, err)
+				if tt.expectedStatus != nil {
+					require.Equal(t, tt.expectedStatus.SecretRef.Name, dialer.SecretRef().Name)
+					require.Equal(t, tt.expectedStatus.SecretRef.Key, dialer.SecretRef().Key)
+					require.Equal(t, tt.expectedStatus.SecretRef.Backend, dialer.SecretRef().Backend)
+					require.Equal(t, "latest-version", dialer.SecretRef().ResourceVersion)
+					require.NotEmpty(t, dialer.SecretRef().ResourceVersion)
+					require.NotEqual(t, "test-version", dialer.SecretRef().ResourceVersion)
+				}
+			}
+		})
+	}
+}
+
+// TestSpecStatusSecretsDifferent tests that the correct result is returned
+// when comparing the secret in the status against the existing secret.
+func TestSpecStatusSecretsDifferent(t *testing.T) {
+	t.Parallel()
+	cases := map[string]struct {
+		dialer      *v1alpha1.PeeringDialer
+		secret      *corev1.Secret
+		isDifferent bool
+	}{
+		"different secret name in spec and status": {
+			dialer: &v1alpha1.PeeringDialer{
+				Spec: v1alpha1.PeeringDialerSpec{
+					Peer: &v1alpha1.Peer{
+						Secret: &v1alpha1.Secret{
+							Name:    "foo",
+							Key:     "token",
+							Backend: "kubernetes",
+						},
+					},
+				},
+				Status: v1alpha1.PeeringDialerStatus{
+					SecretRef: &v1alpha1.SecretRefStatus{
+						Secret: v1alpha1.Secret{
+							Name:    "bar",
+							Key:     "token",
+							Backend: "kubernetes",
+						},
+					},
+				},
+			},
+			secret:      nil,
+			isDifferent: true,
+		},
+		"different secret key in spec and status": {
+			dialer: &v1alpha1.PeeringDialer{
+				Spec: v1alpha1.PeeringDialerSpec{
+					Peer: &v1alpha1.Peer{
+						Secret: &v1alpha1.Secret{
+							Name:    "foo",
+							Key:     "token",
+							Backend: "kubernetes",
+						},
+					},
+				},
+				Status: v1alpha1.PeeringDialerStatus{
+					SecretRef: &v1alpha1.SecretRefStatus{
+						Secret: v1alpha1.Secret{
+							Name:    "foo",
+							Key:     "key",
+							Backend: "kubernetes",
+						},
+					},
+				},
+			},
+			secret:      nil,
+			isDifferent: true,
+		},
+		"different secret backend in spec and status": {
+			dialer: &v1alpha1.PeeringDialer{
+				Spec: v1alpha1.PeeringDialerSpec{
+					Peer: &v1alpha1.Peer{
+						Secret: &v1alpha1.Secret{
+							Name:    "foo",
+							Key:     "token",
+							Backend: "kubernetes",
+						},
+					},
+				},
+				Status: v1alpha1.PeeringDialerStatus{
+					SecretRef: &v1alpha1.SecretRefStatus{
+						Secret: v1alpha1.Secret{
+							Name:    "foo",
+							Key:     "token",
+							Backend: "vault",
+						},
+					},
+				},
+			},
+			secret:      nil,
+			isDifferent: true,
+		},
+		"different secret ref in status and saved secret": {
```

Comment:
Is it possible for the versions to end up out of order, i.e. this version is older than what is stored in the secret?
 if so should we test for that?

Reply:
if the secret changes, it's resource version would change. I don't think there is a way where the spec secret version is older than the one in the status if that makes sense. am I misunderstanding the question?

Reply:
I was just wondering if it were possible to get out of order here due to caching effects or a failed secret update, etc. It's more of a question about how the resourceVersion is compared I suppose. But sounds like this isnt something we'd need to worry about!

Reply:
yeah! i think that aspect should be safe. we will never be in a situation where the status "gets ahead" of the spec. If the status is not updated for whatever reason, re-running the steps will be a no-op and the status will then get updated.